### ISSUE TITLE
Different results. 

### ISSUE THREAD

**Issue by hristorv** [jonrei/tf-AdaIN#1]
Hi there,
I have the code and it doesnt seem to produce the same results as in the paper. What might be different and where is the issue ?

**Comment by jonrei** [comment]
Can you paste the code you are trying to run, and with what input? 
It is hard to say without knowing your details, but in the original code there are several parameters that can be tweaked differently to create a specific example.

**Comment by hristorv** [comment]
Well i didnt change anything and i got same results as you did in the notebook. So the examples you provided and my test samples compared to the same input/style in the paper didnt seem to be the same. 
What parameters can i change to make it similar to the paper ? Also, loading the same models from the paper with torchfile and using them in tensorflow, does it have any difference ?
Btw have you checked this project:
https://github.com/diggerdu/Vincent

**Comment by jonrei** [comment]
There are a couple of features that I haven't implemented, like the `alpha` parameter, see [test.lua](https://github.com/xunhuang1995/AdaIN-style/blob/master/test.lua#L117) for details. It is possible they tweaked that to make the images better looking.
It is also possible that there are bugs in my torch import code, but it is a bit difficult to verify.

I have not checked that project, will take a look!

**Comment by hristorv** [comment]
My test and your examples on github both had that darker look and the colors were a bit different. I have implemented the alpha parameter and set it to 0 which should result in the same input image without change instead i got this :

![paper](https://cloud.githubusercontent.com/assets/4991283/24833119/354d5e42-1cc8-11e7-86a3-5a7d686d1471.png)

**Comment by jonrei** [comment]
Yes you are right, I tried having input and style as the same image too, which results in a darker image. I will try to look into it.
If you want you can do a pull request for the alpha parameter.

**Comment by hristorv** [comment]
I was trying to fix it those days but with no result, Did you manage eto find out what is causing this ?

**Comment by jonrei** [comment]
No I haven't. I found out that the VGG model expects input in BGR format instead of RGB format. I might push a fix later today, but the images are still off and too dark.

**Comment by ArturoDeza** [comment]
@jonrei I haven't looked into the code yet, but I have a hunch the color problem might have to do with normalizing the feature vectors. There is also a standard image mean-normalization procedure to do before pushing the image through the VGGNet, you might want to check that out. See for example: https://github.com/xunhuang1995/AdaIN-style/blob/master/lib/utils.lua#L6

**Comment by hristorv** [comment]
I cant figure this out guys, any1 got any idea what the problem might be ?

**Comment by jonrei** [comment]
No, just pushed a RGB->BGR fix, but it doesn't change much.
@ArturoDeza Normalization is done [here](https://github.com/jonrei/tf-AdaIN/blob/master/AdaIN.py#L79]).

**Comment by hristorv** [comment]
@jonrei  I am getting better result from changing this line [here](https://github.com/jonrei/tf-AdaIN/blob/master/AdaIN.py#L75) by substracting the means (simply change to -=). But the color is still different. I also noticed that if you dont add the bias from the SpatialConvolution module nothing changes, any idea why is that?

**Comment by jonrei** [comment]
@hristorv Fixed the means, had switched them around :)

**Comment by diggerdu** [comment]
But I think it is no need to do this process if you employ the vgg19 model which xunhuang uesd;

```
the first layer is 
{'biases': array([-103.939, -116.779, -123.68 ]), 'weights': array([[[[   0.]],
 
         [[   0.]],
 
         [[ 255.]]],
 
 
        [[[   0.]],
 
         [[ 255.]],
 
         [[   0.]]],
 
 
        [[[ 255.]],
 
         [[   0.]],
 
         [[   0.]]]])}
```

**Comment by jonrei** [comment]
@diggerdu I had missed that! I have tried with and without offset, but the results are still not exactly like the original example. Have a look at the updated examples [with](https://github.com/jonrei/tf-AdaIN/blob/master/AdaIN_example.ipynb) and [without offset](https://github.com/jonrei/tf-AdaIN/blob/master/AdaIN_example_no_offset.ipynb).

**Comment by diggerdu** [comment]
My results is very similar to yours. I think there is some wrong with tf.nn.moments() or tf.pad.
https://github.com/tensorflow/tensorflow/issues/3290

**Comment by hristorv** [comment]
I have solved the problem. The color values of the images are represented from 0 to 1. However they should be from 0 to 255 after postprocess.
What we need to do is preprocess the images, dividing by 255. And also postprocess, multiplying by 255 and clipping the values.
I am getting the same results as in the original paper and code.

**Comment by diggerdu** [comment]
@hristorv  
But I have do what you say
https://github.com/diggerdu/Vincent/blob/master/beta/vincent.py#L17
https://github.com/diggerdu/Vincent/blob/master/beta/vincent.py#L55

**Comment by hristorv** [comment]
@diggerdu You also need to clip the value to 0-255 in postprocess. What model do you use in  your code ? What results do you get?

**Comment by diggerdu** [comment]
@hristorv 
黄勋 used
https://github.com/xunhuang1995/AdaIN-style/blob/master/models/download_models.sh
My results
http://imgur.com/a/qae6S
I clipped in other script
https://github.com/diggerdu/Vincent/blob/master/deploy/test.py#L23

**Comment by jonrei** [comment]
@hristorv Great work! I have committed a version with the [0,1] normalization.
But I can't really wrap my head around this, first we normalize to [0,1], then we subtract the VGG means which obviously are for images in the range [0,255], and then we do not add the means back when we are done. My guess is that the decoder net is trained in this way and compensates for the strange normalization. If that is true, it would probably be possible to get even better result if the normalization is done differently.

**Comment by hristorv** [comment]
@jonrei The last layer of the decoder adds back the means which are in [0,1] form. I want to further improve the algorithm to make it more like the other implementations that have way better quality, but i got no idea where to start. How do you think the normalization should be changed, for better results?

**Comment by jonrei** [comment]
@hristorv First, you need a working training environment, using the MSCOCO dataset.
Set the decoder part to be trainable and the VGG part to not trainable. Use the loss functions from Eq. 12 and 13 in Huangs paper. When that part is done, it is possible to start to experiment with the normalization.
As I understand it, the data to the original VGG19 is only centered around zero, not normalized in any way. So I would try to feed the net with images without the normalization when feeding the VGG net, and try to relearn the decoder net to reconstruct images from that.
If you get the training environment running you could also experiment with a decoder of different design.
